beacon/light/types/committee.go

```diff
@@ -0,0 +1,204 @@
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package types
+
+import (
+	"encoding/json"
+	"errors"
+	"math/bits"
+
+	"github.com/ethereum/go-ethereum/beacon/params"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/minio/sha256-simd"
+	bls "github.com/protolambda/bls12-381-util"
+)
+
+const SerializedCommitteeSize = (params.SyncCommitteeSize + 1) * params.BlsPubkeySize
+
+type SerializedCommittee [SerializedCommitteeSize]byte
+
+// jsonSyncCommittee is the JSON representation of a sync committee
+//
+// See data structure definition here:
+// https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/beacon-chain.md#syncaggregate
+type jsonSyncCommittee struct {
+	Pubkeys   []hexutil.Bytes `json:"pubkeys"`
+	Aggregate hexutil.Bytes   `json:"aggregate_pubkey"`
+}
+
+// MarshalJSON marshals as JSON.
+func (s *SerializedCommittee) MarshalJSON() ([]byte, error) {
+	sc := jsonSyncCommittee{Pubkeys: make([]hexutil.Bytes, params.SyncCommitteeSize)}
+	for i := range sc.Pubkeys {
+		sc.Pubkeys[i] = make(hexutil.Bytes, params.BlsPubkeySize)
+		copy(sc.Pubkeys[i][:], s[i*params.BlsPubkeySize:(i+1)*params.BlsPubkeySize])
+	}
+	sc.Aggregate = make(hexutil.Bytes, params.BlsPubkeySize)
+	copy(sc.Aggregate[:], s[params.SyncCommitteeSize*params.BlsPubkeySize:])
+	return json.Marshal(&sc)
+}
+
+// UnmarshalJSON unmarshals from JSON.
+func (s *SerializedCommittee) UnmarshalJSON(input []byte) error {
+	var sc jsonSyncCommittee
+	if err := json.Unmarshal(input, &sc); err != nil {
+		return err
+	}
+	if len(sc.Pubkeys) != params.SyncCommitteeSize {
+		return errors.New("Invalid number of pubkeys")
+	}
+	for i, key := range sc.Pubkeys {
+		if len(key) != params.BlsPubkeySize {
+			return errors.New("Invalid pubkey size")
+		}
+		copy(s[i*params.BlsPubkeySize:(i+1)*params.BlsPubkeySize], key[:])
+	}
+	if len(sc.Aggregate) != params.BlsPubkeySize {
+		return errors.New("Invalid pubkey size")
+	}
+	copy(s[params.SyncCommitteeSize*params.BlsPubkeySize:], sc.Aggregate[:])
+	return nil
+}
+
+// SerializedCommitteeRoot calculates the root hash of the binary tree representation
+// of a sync committee provided in serialized format
+func (s *SerializedCommittee) Root() common.Hash {
```

Comment:
Basically yes :) I wrote some simple tools for binary merkle trees because I needed kinda special things but then I started using it for hashing everything :) Now I started using tools from [github.com/protolambda/zrnt ](https://github.com/protolambda/zrnt) for hashing beacon blocks, I'll check if I can easily use it here too.